proposals/pubsub-mq-interfaces-proposal.md

```diff
@@ -0,0 +1,144 @@
+
+![](https://i.imgur.com/Rsa310j.png)
+
+###  2.2. <a name='pubsubBackground'></a>`pubsub` Background
+
+`pubsub` is a messaging pattern, where, instead of sending messages to specific receivers, publishers (i.e., message transmitters) send messages to a channel (or, topic). On the other end, subscribers (i.e., message receivers) can own a subscription to that channel, so they will receive whatever message is sent to it.
+
+![](https://i.imgur.com/pC6SIC0.png)
+
+In addition, there are two ways to receive messages:
+1. Pull-based: The subscriber pulls messages from the channel.
+1. Push-based: The channel pushes messages to the subscriber.
+
+Plus, `pubsub`s usually surface the idea of "Quality of Service" (QoS):
+1. At most once: The message is delivered at most once, but could be lost.
+1. At least once: The message delivery is guaranteed, but could be duplicated.
+1. Exactly once: The message *pretty much* delivered once.
+
+###  2.3. <a name='pubsubmessage-queueComparison'></a>`pubsub`/`message-queue` Comparison
+
+Historically, `pubsub` services (i.e., implementers of the `pubsub` pattern)  are thought to be comprised of non-competitive consumers – That is, if a message is sent to a channel, every receiver that is subscribed to that channel will, eventually, receive it.
+
+While, on the other hand, historically, `message-queue`s are thought to be comprised of competitive consumers – That is, a queued message is dequeued as soon as any one receiver has read it.
+
+However, if we look at services that provide `message-queue` / `pubsub` implementations, the line between the two concepts becomes blurred.
+
+That is, today's most used related services provide both `message-queue`, and `pubsub` capabilities. Below, is a visual example from [HiveMQ](https://www.hivemq.com/) displaying an event-driven architecture with both `message-queue`, and `pubsub` capabilities:
+
+![](./images/pubsub-interface-proposal-0.gif)
+Source: [HiveMQ's Shared Subscriptions](https://www.hivemq.com/docs/hivemq/4.9/user-guide/shared-subscriptions.html)
+
+##  3. <a name='pubsubhttpInteroptability'></a>`pubsub`/`http` Interoptability
+
+`pubsub`s are often used in conjuction with `http` to send, and receive messages from subscribed topics. However, through experimentation on the intereoperability between the two services, it is impossible to use `slight`'s `http` service to correctly manage `pubsub` subscriptions. This is because the `http` interface cannot return a handle to the `pubsub` client under which the subscription was created - Hence, you can only receive messages from a subscribed topic if the subscriptions is done in conjunction with the message polling. This is a problem because message polling, and subscribing should be two separate concerns as, if you only subscribe to a topic at the time of polling, you will miss messages that were sent to the topic before you subscribed to it.
```

Comment:
Probably should mention that the underlying limitation is that each http handler is an independent wasm module, meaning there is no global variables to share among different http handlers. Thus, there is no global handle to message broker in the http handler.

Reply:
done ✅